crates/biome_js_analyze/src/lint/nursery/no_restricted_types.rs

```diff
@@ -0,0 +1,136 @@
+};
+use biome_console::markup;
+use biome_deserialize_macros::Deserializable;
+use biome_js_factory::make;
+use biome_js_syntax::{JsReferenceIdentifier, TextRange, TsReferenceType};
+use biome_rowan::BatchMutationExt;
+use biome_unicode_table::is_js_ident;
+use rustc_hash::FxHashMap;
+
+#[cfg(feature = "schemars")]
+use schemars::JsonSchema;
+
+declare_lint_rule! {
+    /// Disallow user defined types.
+    ///
+    /// This rule allows you to specify type names that you don’t want to use in your application.
+    ///
+    /// ## Options
+    ///
+    /// Use the options to specify additional types that you want to restrict in your
+    /// source code.
+    ///
+    /// ```json
+    /// {
+    ///     "//": "...",
+    ///     "options": {
+    ///         "types": {
+    ///            "Foo": {
+    ///               "message": "Only bar is allowed",
+    ///               "use": "bar"
+    ///             },
+    ///             "OldAPI": {
+    ///                 "message": "Use NewAPI instead"
+    ///             }
+    ///         }
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// In the example above, the rule will emit a diagnostics if tried to use `Foo` or `OldAPI` are used.
+    ///
+    pub NoRestrictedTypes {
+        version: "next",
+        name: "noRestrictedTypes",
+        language: "ts",
+        sources: &[
+            RuleSource::EslintTypeScript("no-restricted-types"),
+        ],
+        recommended: false,
+        fix_kind: FixKind::Safe,
+
+    }
+}
+
+impl Rule for NoRestrictedTypes {
+    type Query = Ast<TsReferenceType>;
+    type State = State;
+    type Signals = Option<Self::State>;
+    type Options = NoRestrictedTypesOptions;
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let ts_reference_type = ctx.query();
+        let options = ctx.options();
+
+        let ts_any_name = ts_reference_type.name().ok()?;
+        let identifier = ts_any_name.as_js_reference_identifier()?;
+        let identifier_token = identifier.value_token().ok()?;
+        let token_name = identifier_token.text_trimmed();
+
+        let restricted_type = options.types.get(token_name)?.clone();
+
+        Some(State {
+            restricted_type,
+            range: identifier_token.text_trimmed_range(),
+            identifier: identifier.clone(),
+        })
+    }
+
+    fn diagnostic(_ctx: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> {
+        Some(RuleDiagnostic::new(
+            rule_category!(),
+            state.range,
+            markup! { {state.restricted_type.message} }.to_owned(),
+        ))
+    }
+
+    fn action(ctx: &RuleContext<Self>, state: &Self::State) -> Option<JsRuleAction> {
+        let suggested_type = &state.restricted_type.use_instead.clone()?;
+        if !is_js_ident(suggested_type) {
+            return None;
+        }
+
+        let mut mutation = ctx.root().begin();
+
+        let prev_token = state.identifier.clone().value_token().ok()?;
```

Comment:
I could simply remove `identifier` from `State`, and retrieve again the identifier here:

```suggestion
        let ts_reference_type = ctx.query();
        let ts_any_name = ts_reference_type.name().ok()?;
        let identifier = ts_any_name.as_js_reference_identifier()?;
        let prev_token = identifier.value_token().ok()?;
```

Also, destructing the reference instead of calling `as_js_reference_identifier()` avoids cloning the node here.